ExampleMod/Content/Projectiles/ExampleLaser.cs

```diff
@@ -0,0 +1,252 @@
+			// Set a custom immune time of 5 frames upon hitting something.
+			target.immune[Owner.whoAmI] = 5;
+		}
+
+		private void UpdatePlayer() {
+			// Only change the projectiles variables if the current player is the owner of the projectile.
+			if (Owner.whoAmI == Main.myPlayer) {
+				// Get the unit vector direction to the players mouse, and set the projectile's velocity to it.
+				Vector2 aimDirection = Owner.Center.DirectionTo(Main.MouseWorld);
+				Projectile.velocity = aimDirection;
+				Projectile.direction = Main.MouseWorld.X > Owner.Center.X ? 1 : -1;
+			}
+
+			// Make the player's direction match the projectile's.
+			Owner.ChangeDir(Projectile.direction);
+			// Set this projectile as the player's held projectile.
+			Owner.heldProj = Projectile.whoAmI;
+			// Set the item time and animation length to 2 frames while the projectile is in use.
+			Owner.itemTime = Owner.itemAnimation = 2;
+			// Update the item rotation to the aim direction, accounting for flipping the rotation if facing to the left.
+			Owner.itemRotation = Projectile.velocity.ToRotation() + (Projectile.direction < 0 ? MathHelper.Pi : 0f);
+		}
+
+		private void ChargeLaser() {
+			// Increment the charge amount.
+			if (LaserCharge < MaxCharge) {
+				LaserCharge++;
+			}
+
+			// Get the position to spawn the charging dust at.
+			Vector2 offset = Projectile.velocity * (HoldOffset - 20f);
+			Vector2 baseDustPosition = Owner.Center + offset - Vector2.One * 10f;
+
+			// Get the amount of dust to spawn per frame, this increases with charge.
+			int amountOfDust = (int)(LaserCharge / 30f);
+
+			for (int i = 0; i < amountOfDust; i++) {
+				// Offset the spawn position of the dust randomly, getting closer to the original point over the charge length.
+				Vector2 actualDustPosition = baseDustPosition + Main.rand.NextVector2Unit() * (12f - amountOfDust * 2f);
+				// Get the velocity for the dust based on the offset position, getting slower over the charge length.
+				Vector2 dustVelocity = baseDustPosition.DirectionTo(actualDustPosition) * 1.5f * (10f - amountOfDust * 2f) / 10f;
+				float dustScale = Main.rand.NextFloat(0.5f, 1f);
+
+				Dust dust = Dust.NewDustDirect(actualDustPosition, 20, 20, DustID.Electric, dustVelocity.X, dustVelocity.Y, Scale: dustScale);
+				// Disable gravity for the dust.
+				dust.noGravity = true;
+			}
+		}
+
+		private void SetLaserPosition() {
+			float chargeIncrement = 5f;
+			// Loop along the laser's distance to determine the maximum distance it can hit.
+			for (LaserDistance = HoldOffset; LaserDistance <= MaxDistance; LaserDistance += chargeIncrement) {
+				Vector2 checkPosition = Owner.Center + Projectile.velocity * LaserDistance;
+
+				// If the current checkPosition is blocked via collision, decrease the distance to the last good value and break out of the loop.
+				if (!Collision.CanHit(Owner.Center, Projectile.width, Projectile.height, checkPosition, Projectile.width, Projectile.height)) {
```

Comment:
This needs to be changed to `Collision.CanHitLine()` as pointed out by this video (3 years ago lol) https://youtu.be/ZSi0bo7kZmM

Reply:
Has been updated, thank you :)